lib/Serialization/SerializeDoc.cpp

```diff
@@ -502,7 +502,390 @@ void serialization::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
 
   S.writeToStream(os);
 }
+namespace {
+struct LineColoumn {
+  uint32_t Line = UINT32_MAX;
+  uint32_t Column = UINT32_MAX;
+};
+
+struct DeclLocationsTableData {
+  uint32_t SourceFileID;
+  LineColoumn Loc;
+  LineColoumn NameLoc;
+  LineColoumn StartLoc;
+  LineColoumn EndLoc;
+};
+
+class DeclLocsTableInfo {
+public:
+  using key_type = uint32_t;
+  using key_type_ref = key_type;
+  using data_type = DeclLocationsTableData;
+  using data_type_ref = const data_type &;
+  using hash_value_type = uint32_t;
+  using offset_type = unsigned;
+
+  hash_value_type ComputeHash(key_type_ref key) {
+    return key;
+  }
+
+  std::pair<unsigned, unsigned>
+  EmitKeyDataLength(raw_ostream &out, key_type_ref key, data_type_ref data) {
+    const unsigned numLen = 4;
+    const unsigned LineColumnLen = numLen * 2;
+    uint32_t keyLength = numLen;
+    uint32_t dataLength = 0;
+
+    // File ID
+    dataLength += numLen;
+
+    // Loc
+    dataLength += LineColumnLen;
+
+    // NameLoc
+    dataLength += LineColumnLen;
+
+    // StartLoc
+    dataLength += LineColumnLen;
+
+    // EndLoc
+    dataLength += LineColumnLen;
+    endian::Writer writer(out, little);
+    writer.write<uint32_t>(keyLength);
```

Comment:
If the length is constant you actually don't have to serialize it at all; returning it is sufficient.